main/src/library/Chain.flix

```diff
@@ -0,0 +1,847 @@
+/*
+ * Copyright 2020
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+///
+/// The Chain type.
+///
+/// A chain is a list represented as an unbalanced binary tree.
+/// It supports efficient append and "snoc" - appending elements at the tail
+/// of the list.
+///
+/// Note - the constructors `Empty`, `One` and `Chain` should not be used directly.
```

Comment:
How does Haskell do it? Is the idea that you can export a type without exporting its constructors?

Reply:
In the modules export list, to export a type with constructors you can do `Chain(..)` to export all the constructors or e.g. `Chain(Empty)` to export the type and only the named constructors.

To export a type and hide all the constructors just list the type name in the module export with no parens / constructor list.

Reply:
OK, thanks. We should think about that.